main/src/ca/uwaterloo/flix/language/phase/Lowering.scala

```diff
@@ -538,27 +555,73 @@ object Lowering {
       val ms = methods.map(visitJvmMethod)
       Expression.NewObject(name, clazz, t, pur, eff, ms, loc)
 
+    // New channel expressions are rewritten as follows:
+    //     chan Int32 10
+    // becomes:
+    //     newChannel[Int32](10)
+    //
     case Expression.NewChannel(exp, tpe, pur, eff, loc) =>
       val e = visitExp(exp)
       val t = visitType(tpe)
-      Expression.NewChannel(e, t, pur, eff, loc)
+      val newChannel = Expression.Def(Defs.ChannelNew, Type.mkImpureArrow(e.tpe, t, loc), loc)
+      Expression.Apply(newChannel, e :: Nil, t, pur, eff, loc)
 
+    // Channel get expressions are rewritten as follows:
+    //     <- c
+    // becomes:
+    //     get(c)
     case Expression.GetChannel(exp, tpe, pur, eff, loc) =>
       val e = visitExp(exp)
       val t = visitType(tpe)
-      Expression.GetChannel(e, t, pur, eff, loc)
-
-    case Expression.PutChannel(exp1, exp2, tpe, pur, eff, loc) =>
+      val getChannel = Expression.Def(Defs.ChannelGet, Type.mkImpureArrow(e.tpe, t, loc), loc)
+      Expression.Apply(getChannel, e :: Nil, t, pur, eff, loc)
+
+    // Channel put expressions are rewritten as follows:
+    //     c <- 42
+    // becomes:
+    //     put(42, c)
+    case Expression.PutChannel(exp1, exp2, _, pur, eff, loc) =>
       val e1 = visitExp(exp1)
       val e2 = visitExp(exp2)
-      val t = visitType(tpe)
-      Expression.PutChannel(e1, e2, t, pur, eff, loc)
-
+      val putChannel = Expression.Def(Defs.ChannelPut, Type.mkImpureUncurriedArrow(List(e2.tpe, e1.tpe), Type.Unit, loc), loc)
+      Expression.Apply(putChannel, List(e2, e1), Type.Unit, pur, eff, loc)
+
+    // Channel select expressions are rewritten as follows:
+    //     select {
+    //         case x <- ?ch1 => ?handlech1
+    //         case y <- ?ch2 => ?handlech2
+    //         case _ => ?default
+    //     }
+    // becomes:
+    //     let ch1 = ?ch1;
+    //     let ch2 = ?ch2;
+    //     match selectFrom([mpmcAdmin(ch1), mpmcAdmin(ch2)]) @ Static, false) {  // true if no default
+    //         case (0, locks) =>
+    //             let x = unsafeGetAndUnlock(ch1, locks);
+    //             ?handlech1
+    //         case (1, locks) =>
+    //             let y = unsafeGetAndUnlock(ch2, locks);
+    //             ?handlech2
+    //         case (-1, _) =>                                                  // omitted if no default
+    //             ?default                                                     //
+    //         case _ =>
+    //             unreachable!()
+    //     }
     case Expression.SelectChannel(rules, default, tpe, pur, eff, loc) =>
       val rs = rules.map(visitSelectChannelRule)
       val d = default.map(visitExp)
       val t = visitType(tpe)
-      Expression.SelectChannel(rs, d, t, pur, eff, loc)
+
+      val channels = (rs map { case SelectChannelRule(_, c, _) => (mkLetSym("chan", loc), c) }).toArray
```

Comment:
We tend to never use arrays because they are mutable. Is there a particular reason you need an array over a List or Seq?

Reply:
I think its just lefter code in channels. It could be changed in Channels.flix

Reply:
It's because I wanted to access it by numeric index, i.e. channels(i).

I could have achieved the same by zipping (as per later comment) but thought that this was simpler/easier to understand.